src/gui/userstatus.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) by Camila <hello@camila.codes>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ */
+
+#include "userstatus.h"
+#include "account.h"
+#include "accountstate.h"
+#include "networkjobs.h"
+#include "folderman.h"
+#include "creds/abstractcredentials.h"
+#include "theme.h"
+
+#include <QTimer>
+#include <QJsonDocument>
+#include <QJsonObject>
+
+namespace OCC {
+
+UserStatus::UserStatus(QObject *parent)
+    : QObject(parent)
+    , _status(Status::Online)
+    , _message("")
+{
+}
+
+UserStatus::Status UserStatus::stringToEnum(const QString &status) const 
+{
+    // api should return invisible, dnd,... toLower() it is to make sure 
+    // it matches _preDefinedStatus, otherwise the default is online (0)
+    const auto statusEnum = _preDefinedStatus.value(status.toLower(), 0);
+    return static_cast<Status>(statusEnum);
+}
+
+void UserStatus::fetchUserStatus(AccountPtr account)
+{
+    if (_job) {
+        _job->deleteLater();
+    }
+
+    _job = new JsonApiJob(account, QStringLiteral("/ocs/v2.php/apps/user_status/api/v1/user_status"), this);
+    connect(_job.data(), &JsonApiJob::jsonReceived, this, &UserStatus::slotFetchUserStatusFinished);
+    _job->start();
+}
+
+void UserStatus::slotFetchUserStatusFinished(const QJsonDocument &json)
+{
+    const auto retrievedData = json.object().value("ocs").toObject().value("data").toObject();
+    const auto emoji = retrievedData.value("icon").toString();
+    const auto message = retrievedData.value("message").toString();
+    auto statusString = retrievedData.value("status").toString(); 
+    _status = stringToEnum(statusString);
+    
+    // to display it to the user like 'Invisible' instead of 'invisible'
+    statusString.replace(0, 1, statusString.at(0).toUpper());  
```

Comment:
@camilasan Can we end up having an empty `statusString`, I am guessing, this will crash then? Maybe we need to check for `!isEmpty()`?

Reply:
theoretically it will never be empty, but it doesn't hurt to be careful :)

Reply:
@camilasan Actually, I tried running this branch and got an assert exactly at this place with an empty string :) 

2021-03-24 14:47:20:595 [ fatal default ]:	ASSERT: "uint(i) < uint(size())" in file c:\users\qt\work\qt\qtbase\src\corelib\tools\qstring.h, line 936